arch/x86/core/userspace.S

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel_structs.h>
+#include <arch/x86/asm.h>
+#include <offsets_short.h>
+
+GTEXT(_x86_syscall_invoke)
+GTEXT(_x86_syscall_entry_stub)
+GTEXT(_x86_syscall_entry)
+GTEXT(_x86_userspace_enter)
+
+/* Landing site for syscall SW IRQ. Marshal arguments and call C function for
+ * further processing.
+ */
+SECTION_FUNC(TEXT, _x86_syscall_entry_stub)
+	push	%esi		/* arg5 */
+	push	%edi		/* arg4 */
+	push	%ebx		/* arg3 */
+#ifndef CONFIG_X86_IAMCU
+	push	%ecx		/* arg2 */
+	push	%edx		/* arg1	*/
+	push	%eax		/* Call id */
+#endif
+
+	call	_x86_syscall_entry
+
+	/* EAX now contains return value. Pop or xor everything else to prevent
+	 * information leak from kernel mode.
+	 */
+#ifndef CONFIG_X86_IAMCU
+	add	$4, %esp	/* old EAX value */
+	pop	%edx
+	pop	%ecx
+#else
+	xor	%edx, %edx
+	xor	%ecx, %ecx
+#endif
+	pop	%ebx
+	pop	%edi
+	pop	%esi
+
+	iret
+
+
+/* FUNC_NORETURN void _x86_userspace_enter(_thread_entry_t user_entry,
+ *					   void *p1, void *p2, void *p3,
+ *					   u32_t stack_end,
+ *					   u32_t stack_start)
+ *
+ * A one-way trip to userspace.
+ */
+SECTION_FUNC(TEXT, _x86_userspace_enter)
+	add	$0x4, %esp /* Discard return address on stack */
```

Comment:
Hah, here's a spot where the reverse is true: make this a POP instruction instead (looks like %esi is going to be clobbered immediately below and is a good choice).   Much smaller instruction than the add, which the assembler (probably) won't be able to optimize because of the side effect from POP (it's smart, but not that smart).

Reply:
good idea